sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -124,12 +124,45 @@ case class GroupExpression(children: Seq[Expression]) extends Expression {
   override def dataType: DataType = throw new UnsupportedOperationException
 }
 
+trait TypeConstraint {
+  self: Expression =>
+
+  def isTypeMatch: Boolean
+
+  def errorMessage: String
+
+  override lazy val resolved = childrenResolved && isTypeMatch
+}
+
 /**
  * Expressions that require a specific `DataType` as input should implement this trait
  * so that the proper type conversions can be performed in the analyzer.
  */
-trait ExpectsInputTypes {
+trait ExpectsInputTypes extends TypeConstraint {
+  self: Expression =>
 
   def expectedChildTypes: Seq[DataType]
 
+  def isTypeMatch: Boolean = children.map(_.dataType) == expectedChildTypes
+
+  // This won't be called as we always do type cast for it in `HiveTypeCoercion`.
+  def errorMessage: String = throw new UnsupportedOperationException
+}
+
+trait TypeEqualConstraint extends TypeConstraint {
```

Comment:
I'd remove this and just put it in the 3 places that use this right now. It is not that big of deal to duplicate this tiny amount of code.